lib/THCUNN/SparseLinear.cu

```diff
@@ -0,0 +1,258 @@
+#include "THCUNN.h"
+#include <cusparse.h>
+#include <thrust/device_vector.h>
+
+static cusparseHandle_t cusparse_handle = 0;
+
+static void init_cusparse() {
+  if (cusparse_handle == 0) {
+    cusparseStatus_t status = cusparseCreate(&cusparse_handle);
+    if (status != CUSPARSE_STATUS_SUCCESS) {
+      THError("CUSPARSE Library initialization failed");
+    }
+  }
+}
+
+static bool checkInput(THCudaTensor* t)
+{
+  return t->nDimension == 2 && t->size[1] == 3;
+}
+
+static bool checkSize2D(THCudaTensor* t, long size0, long size1)
+{
+  return t->nDimension == 2 && t->size[0] == size0 && t->size[1] == size1;
+}
+
+static bool checkSize1D(THCudaTensor* t, long size0)
+{
+  return t->nDimension == 1 && t->size[0] == size0;
+}
+
+void printCuda(THCState *state, THCudaTensor *tensor, char* str) {
+  thrust::device_ptr<float> ptr(THCudaTensor_data(state, tensor));
+  printf("%s\n", str);
+  thrust::copy(ptr, ptr+THCudaTensor_nElement(state, tensor), std::ostream_iterator<float>(std::cout, "\t"));
+  printf("\n");
+}
+void printCuda(THCState *state, THCudaIntTensor *tensor, char* str) {
+  thrust::device_ptr<int> ptr(THCudaIntTensor_data(state, tensor));
+  printf("%s\n", str);
+  thrust::copy(ptr, ptr+THCudaIntTensor_nElement(state, tensor), std::ostream_iterator<float>(std::cout, "\t"));
+  printf("\n");
+}
+
+static THCudaTensor *values = 0;
```

Comment:
this seems like a terrible idea (and it's also not thread-safe btw), let's dicuss offline why you're trying to keep state in this file, i think you can remove it.